model/DeliveryContainerService.php

```diff
@@ -177,4 +169,44 @@ protected function getActiveFeatures(\core_kernel_classes_Resource $delivery)
             $delivery->getOnePropertyValue($this->getProperty(self::TEST_RUNNER_FEATURES_PROPERTY))
         );
     }
+
+    /**
+     * @param core_kernel_classes_Resource $delivery
+     * @return array
+     */
+    protected function getPluginsDisabledForDelivery(core_kernel_classes_Resource $delivery)
+    {
+        $enabledPlugins = [];
+        $disabledPlugins = [];
+        try {
+            $testRunnerFeatureService = $this->getServiceLocator()->get(TestRunnerFeatureService::SERVICE_ID);
+            $allTestRunnerFeatures = $testRunnerFeatureService->getAll();
+
+            if (count($allTestRunnerFeatures) === 0) {
+                return $disabledPlugins;
+            }
+
+            $activeTestRunnerFeaturesIds = $this->getActiveFeatures($delivery);
+            foreach ($allTestRunnerFeatures as $feature) {
+                if (in_array($feature->getId(), $activeTestRunnerFeaturesIds)) {
+                    $enabledPlugins = array_merge($enabledPlugins, $feature->getPluginsIds());
+                } else {
+                    $disabledPlugins = array_merge($disabledPlugins, $feature->getPluginsIds());
```

Comment:
`array_merge` in a loop means quadratic complexity.
Consider appending chunks and merging them after the loop. Like
```
<?php
$enabledPluginChunks = [[]];
$disabledPluginChunks = [[]];

foreach ($allTestRunnerFeatures as $feature) {
    if (in_array($feature->getId(), $activeTestRunnerFeaturesIds)) {
        $enabledPluginChunks[] = $feature->getPluginsIds();
    } else {
        $disabledPluginChunks[] = $feature->getPluginsIds();
    }
}

$enabledPlugins = array_merge(...$enabledPluginChunks);
$disabledPlugins = array_merge(...$disabledPluginChunks);
```

Reply:
Or the following, if the app is supposed to run under PHP < 5.6
```php
$enabledPlugins = call_user_func_array('array_merge', $enabledPluginChunks);
$disabledPlugins = call_user_func_array('array_merge', $disabledPluginChunks);
```

Reply:
Fixed